File: parse_courses.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Parse university course listing HTML file line-by-line (memory efficient).
"""

import re
from collections import defaultdict

def main():
    print("Reading file line by line...")
    
    courses = []
    in_script = False
    current_tds = []
    in_td = False
    td_text = ""
    in_thead = False
    
    with open(r'c:\Users\ehdrjs10w\Desktop\mammoth\txt', 'r', encoding='utf-8') as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            
            # Track script blocks
            if '<script' in line:
                in_script = True
                continue
            if '</script>' in line:
                in_script = False
                continue
            if in_script:
                continue
            
            # Track thead
            if '<thead>' in line:
                in_thead = True
                continue
            if '</thead>' in line:
                in_thead = False
                continue
            if in_thead:
                continue
            
            # Skip input, a tags, empty lines
            if '<input ' in line or '<a ' in line or '</a>' in line:
                continue
            
            # New row
            if '<tr>' in line:
                current_tds = []
                continue
            
            # End row - process what we have
            if '</tr>' in line:
                if len(current_tds) >= 11:
                    course = {
                        'code': current_tds[0],
                        'name': current_tds[1],
                        'section': current_tds[2],
                        'credits': current_tds[3],
                        'professor': current_tds[4],
                        'room': current_tds[5],
                        'time': current_tds[6],
                        'semester': current_tds[7],
                        'limit': current_tds[8],
                        'enrolled': current_tds[9],
                        'category': current_tds[10],
                    }
                    if course['code'] and course['name']:
                        courses.append(course)
                current_tds = []
                continue
            
            # Start of td
            if '<td' in line:
                # Check if td content is on same line
                m = re.search(r'<td[^>]*>(.*?)(?:</td>)?', line)
                if m:
                    content = m.group(1).strip()
                    content = re.sub(r'<[^>]+>', '', content).strip()  # strip any inner tags
                    if '</td>' in line:
                        current_tds.append(content)
                        in_td = False
                    else:
                        in_td = True
                        td_text = content
                continue
            
            # End td
            if '</td>' in line:
                in_td = False
                current_tds.append(td_text.strip())
                td_text = ""
                continue
            
            # Content inside td
            if in_td:
                clean = re.sub(r'<[^>]+>', '', line).strip()
                if clean:
                    if td_text:
                        td_text += " " + clean
                    else:
                        td_text = clean
            
            if line_num % 10000 == 0:
                print(f"  Processed {line_num} lines, found {len(courses)} courses so far...")
    
    print(f"\nTotal: {len(courses)} courses found")
    
    if not courses:
        print("No courses found!")
        return
    
    def dw(s):
        """Display width for CJK characters."""
        w = 0
        for ch in s:
            if ord(ch) > 0x1100:
                w += 2
            else:
                w += 1
        return w
    
    def pad(s, tw, align='l'):
        """Pad to target width."""
        d = tw - dw(s)
        if d < 0: d = 0
        if align == 'c':
            return ' ' * (d // 2) + s + ' ' * (d - d // 2)
        elif align == 'r':
            return ' ' * d + s
        return s + ' ' * d
    
    headers = ['교과목코드', '교과목명', '분반', '학점', '교수명', '강의실', '수업시간', '셀학기', '제한인원', '수강인원', '이수구분']
    keys = ['code', 'name', 'section', 'credits', 'professor', 'room', 'time', 'semester', 'limit', 'enrolled', 'category']
    center_keys = {'code', 'section', 'credits', 'limit', 'enrolled', 'semester'}
    
    def make_table(clist):
        # Calculate widths
        ws = []
        for h, k in zip(headers, keys):
            mx = max((dw(c.get(k, '')) for c in clist), default=0)
            ws.append(max(dw(h), mx))
        
        lines = []
        lines.append('┌' + '┬'.join('─' * (w+2) for w in ws) + '┐')
        
        hdr = '│'
        for h, w in zip(headers, ws):
            hdr += ' ' + pad(h, w, 'c') + ' │'
        lines.append(hdr)
        lines.append('├' + '┼'.join('─' * (w+2) for w in ws) + '┤')
        
        for c in clist:
            row = '│'
            for k, w in zip(keys, ws):
                v = c.get(k, '')
                a = 'c' if k in center_keys else 'l'
                row += ' ' + pad(v, w, a) + ' │'
            lines.append(row)
        
        lines.append('└' + '┴'.join('─' * (w+2) for w in ws) + '┘')
        return '\n'.join(lines)

    # Write full list
    p1 = r'c:\Users\ehdrjs10w\Desktop\mammoth\수강목록_전체.txt'
    with open(p1, 'w', encoding='utf-8') as f:
        f.write('═' * 120 + '\n')
        f.write('                         2026학년도 1학기 수강 과목 목록\n')
        f.write('═' * 120 + '\n\n')
        f.write(f'총 {len(courses)}개 교과목\n\n')
        f.write(make_table(courses))
        f.write('\n')
    print(f"Written: {p1}")
    
    # Group by category
    by_cat = defaultdict(list)
    for c in courses:
        cat = c.get('category', '') or '미분류'
        by_cat[cat].append(c)
    
    p2 = r'c:\Users\ehdrjs10w\Desktop\mammoth\수강목록_이수구분별.txt'
    with open(p2, 'w', encoding='utf-8') as f:
        f.write('═' * 120 + '\n')
        f.write('                    2026학년도 1학기 수강 과목 목록 (이수구분별)\n')
        f.write('═' * 120 + '\n\n')
        f.write(f'총 {len(courses)}개 교과목\n\n')
        for cat in sorted(by_cat.keys()):
            cl = by_cat[cat]
            f.write(f'\n▶ {cat} ({len(cl)}개)\n')
            f.write('─' * 80 + '\n')
            f.write(make_table(cl))
            f.write('\n\n')
    print(f"Written: {p2}")
    
    print("\n=== 이수구분별 요약 ===")
    for cat in sorted(by_cat.keys()):
        print(f"  {cat}: {len(by_cat[cat])}개")
    print(f"  Total: {len(courses)}개")


if __name__ == '__main__':
    main()

File: parse_courses.js
const fs = require('fs');

const inputPath = String.raw`c:\Users\ehdrjs10w\Desktop\mammoth\txt`;
const outputPath = String.raw`c:\Users\ehdrjs10w\Desktop\mammoth\수강목록_전체.txt`;
const outputPath2 = String.raw`c:\Users\ehdrjs10w\Desktop\mammoth\수강목록_이수구분별.txt`;

console.log('Reading file...');
let content = fs.readFileSync(inputPath, 'utf-8');
console.log(`File size: ${content.length} chars`);

// Remove script blocks
content = content.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '');

// Remove input tags
content = content.replace(/<input[^>]*\/?>/gi, '');

// Remove <a> tags and content
content = content.replace(/<a[^>]*>[\s\S]*?<\/a>/gi, '');

// Remove thead sections
content = content.replace(/<thead>[\s\S]*?<\/thead>/gi, '');

console.log('Cleaned content. Extracting rows...');

// Find all <tr> rows
const rowRegex = /<tr>([\s\S]*?)<\/tr>/gi;
let rowMatch;
const courses = [];

while ((rowMatch = rowRegex.exec(content)) !== null) {
    const rowContent = rowMatch[1];

    // Extract all <td> contents
    const tdRegex = /<td[^>]*>([\s\S]*?)<\/td>/gi;
    let tdMatch;
    const tds = [];

    while ((tdMatch = tdRegex.exec(rowContent)) !== null) {
        let text = tdMatch[1].replace(/<[^>]+>/g, '').trim();
        tds.push(text);
    }

    if (tds.length < 11) continue;
    if (!tds[0] || !tds[1]) continue;

    courses.push({
        code: tds[0],
        name: tds[1],
        section: tds[2],
        credits: tds[3],
        professor: tds[4],
        room: tds[5],
        time: tds[6],
        semester: tds[7] || '',
        limit: tds[8] || '',
        enrolled: tds[9] || '',
        category: tds[10] || ''
    });
}

console.log(`Found ${courses.length} courses`);

if (courses.length === 0) {
    console.log('No courses found!');
    process.exit(1);
}

// Display width helper for CJK characters
function displayWidth(str) {
    let w = 0;
    for (const ch of str) {
        const code = ch.codePointAt(0);
        if (code >= 0x1100 && code <= 0xFFFF) {
            w += 2;
        } else {
            w += 1;
        }
    }
    return w;
}

function padStr(str, targetWidth, align = 'left') {
    const cur = displayWidth(str);
    const padding = Math.max(0, targetWidth - cur);
    if (align === 'center') {
        const left = Math.floor(padding / 2);
        const right = padding - left;
        return ' '.repeat(left) + str + ' '.repeat(right);
    } else if (align === 'right') {
        return ' '.repeat(padding) + str;
    }
    return str + ' '.repeat(padding);
}

const headers = ['교과목코드', '교과목명', '분반', '학점', '교수명', '강의실', '수업시간', '셀학기', '제한인원', '수강인원', '이수구분'];
const keys = ['code', 'name', 'section', 'credits', 'professor', 'room', 'time', 'semester', 'limit', 'enrolled', 'category'];
const centerKeys = new Set(['code', 'section', 'credits', 'limit', 'enrolled', 'semester']);

function buildTable(courseList) {
    // Calculate column widths
    const widths = headers.map((h, i) => {
        const key = keys[i];
        const hw = displayWidth(h);
        const maxVal = Math.max(...courseList.map(c => displayWidth(c[key] || '')), 0);
        return Math.max(hw, maxVal);
    });

    const lines = [];

    // Top border
    lines.push('┌' + widths.map(w => '─'.repeat(w + 2)).join('┬') + '┐');

    // Header row
    let hdr = '│';
    headers.forEach((h, i) => {
        hdr += ' ' + padStr(h, widths[i], 'center') + ' │';
    });
    lines.push(hdr);

    // Separator
    lines.push('├' + widths.map(w => '─'.repeat(w + 2)).join('┼') + '┤');

    // Data rows
    for (const c of courseList) {
        let row = '│';
        keys.forEach((k, i) => {
            const val = c[k] || '';
            const align = centerKeys.has(k) ? 'center' : 'left';
            row += ' ' + padStr(val, widths[i], align) + ' │';
        });
        lines.push(row);
    }

    // Bottom border
    lines.push('└' + widths.map(w => '─'.repeat(w + 2)).join('┴') + '┘');

    return lines.join('\n');
}

// Write full list
let out1 = '';
out1 += '═'.repeat(120) + '\n';
out1 += '                         2026학년도 1학기 수강 과목 목록\n';
out1 += '═'.repeat(120) + '\n\n';
out1 += `총 ${courses.length}개 교과목\n\n`;
out1 += buildTable(courses) + '\n';

fs.writeFileSync(outputPath, out1, 'utf-8');
console.log(`Written: ${outputPath} (${courses.length} courses)`);

// Group by category
const byCategory = {};
for (const c of courses) {
    const cat = c.category || '미분류';
    if (!byCategory[cat]) byCategory[cat] = [];
    byCategory[cat].push(c);
}

const catNames = Object.keys(byCategory).sort();

let out2 = '';
out2 += '═'.repeat(120) + '\n';
out2 += '                    2026학년도 1학기 수강 과목 목록 (이수구분별)\n';
out2 += '═'.repeat(120) + '\n\n';
out2 += `총 ${courses.length}개 교과목\n\n`;

for (const cat of catNames) {
    const cl = byCategory[cat];
    out2 += `\n▶ ${cat} (${cl.length}개)\n`;
    out2 += '─'.repeat(80) + '\n';
    out2 += buildTable(cl) + '\n\n';
}

fs.writeFileSync(outputPath2, out2, 'utf-8');
console.log(`Written: ${outputPath2}`);

console.log('\n=== 이수구분별 요약 ===');
for (const cat of catNames) {
    console.log(`  ${cat}: ${byCategory[cat].length}개`);
}
console.log(`  Total: ${courses.length}개`);
